### DISCUSSION TITLE
Weird output from fish when using with pexpect

### DISCUSSION THREAD

**Discussion by rsyring** [fish-shell/fish-shell#10747] [Q&A] [answered]
I have a CLI app that I'm writing integration tests for.  I have basically the same integration test for both Bash and Fish and I'm using pexpect to run the tests.  I see Fish has it's own pexpect tests, and the bash tests worked as expect, so I thought this would be straight forward.

Unfortunately I get weird output from fish when using it with pexpect that appears to be some kind of echo.  Any idea how I can turn it off?

```python
    def test_mcve(self):
        env = {'TERM': 'dumb'}
        fish = pexpect.spawn(
            'fish',
            # Options are to avoid contamination from dev's init files and keep tests
            # consistent across hosts.
            ['--no-config', '--debug-output=/tmp/fish-debug.txt'],
            echo=False,
            encoding='utf-8',
            timeout=2,
            env=env,
        )
        fish.sendline('echo foo')
        fish.expect('foo')
        assert fish.before == ''
```

Gives the output:

```
FAILED test_shells.py::TestFish::test_mcve - AssertionError: assert '\rrsyring@meld ~/p/e/s/e/tests (main)> \rrsyring@meld ~/p/e/s/e/tests (main)> echo \rrsyring@meld ~/p/e/s/e/tests (main)> echo ' == ''
  
  + 
  + rsyring@meld ~/p/e/s/e/tests (main)> 
  + rsyring@meld ~/p/e/s/e/tests (main)> echo 
  + rsyring@meld ~/p/e/s/e/tests (main)> echo
```

I understand why there would be a single prompt line.  But why three of them and why do the last two add the "echo" command?

I took a look at Fish's [`pexect_helper.py`](https://github.com/fish-shell/fish-shell/blob/master/build_tools/pexpect_helper.py) to see if they did some kind of special setup with fish or pexpect but I don't see anything.

Thanks in advance.

**Comment by faho** [comment]
Fish does a *lot* of drawing to enable syntax highlighting and autosuggestions.

In your specific case since you set TERM=dumb most of it is useless, but if you had a normal terminal, I guess what happens is along the lines of:

- the prompt is printed
- the command is added, in red (because "ech" isn't a valid command)
- (suggestions are added and removed as necessary)
- the command is found and redrawn in command color

There are some cases where this *could* be optimized to do less, but that needs to be done with care - e.g. if the terminal disagrees on the width of some character, fish should do more explicit cursor movement to avoid overdrawing the line, which leads to a weird staircase-effect that renders the shell almost unusable.

(it also barely matters given that this is a few bytes and computers have gotten really really really fast in the last 30 years)

**Comment by rsyring** [comment]
Thanks for the explanation.  Is there any way to turn all that off and really make it act dumb?  I.e. not echo anything simply present the prompt, take input, give output from the command?  

It's not the speed that concerns me here, just purity and expectations for automated integration tests.

**Reply by faho** [reply, marked as answer]
>  Is there any way to turn all that off and really make it act dumb?

There is not, no.
